src/client/Microsoft.Identity.Client/Internal/Broker/BrokerSilentRequest.cs

```diff
@@ -0,0 +1,102 @@
+        }
+
+        private async Task<MsalTokenResponse> SendAndVerifyResponseAsync()
+        {
+            CreateRequestParametersForBroker();
+
+            MsalTokenResponse msalTokenResponse =
+                await Broker.AcquireTokenUsingBrokerAsync(BrokerPayload).ConfigureAwait(false);
+
+            ValidateResponseFromBroker(msalTokenResponse);
+            return msalTokenResponse;
+        }
+
+        internal void CreateRequestParametersForBroker()
+        {
+            BrokerPayload.Clear();
+            BrokerPayload.Add(BrokerParameter.Authority, _authenticationRequestParameters.Authority.AuthorityInfo.CanonicalAuthority);
+            string scopes = EnumerableExtensions.AsSingleString(_authenticationRequestParameters.Scope);
+
+            BrokerPayload.Add(BrokerParameter.Scope, scopes);
+            BrokerPayload.Add(BrokerParameter.ClientId, _authenticationRequestParameters.ClientId);
+            BrokerPayload.Add(BrokerParameter.CorrelationId, _authenticationRequestParameters.RequestContext.Logger.CorrelationId.ToString());
+            BrokerPayload.Add(BrokerParameter.ClientVersion, MsalIdHelper.GetMsalVersion());
+            BrokerPayload.Add(BrokerParameter.RedirectUri, _serviceBundle.Config.RedirectUri);
+
+            string extraQP = string.Join("&", _authenticationRequestParameters.ExtraQueryParameters.Select(x => x.Key + "=" + x.Value));
+            BrokerPayload.Add(BrokerParameter.ExtraQp, extraQP);
+
+            BrokerPayload.Add(BrokerParameter.Username, _authenticationRequestParameters.Account?.Username ?? string.Empty);
+            BrokerPayload.Add(BrokerParameter.ExtraOidcScopes, BrokerParameter.OidcScopesValue);
+            BrokerPayload.Add(BrokerParameter.HomeAccountId, _silentParameters.Account.HomeAccountId.Identifier);
+            BrokerPayload.Add(BrokerParameter.LocalAccountId, _silentParameters.Account.HomeAccountId.TenantId);
+            BrokerPayload.Add(BrokerParameter.ForceRefresh, _silentParameters.ForceRefresh.ToString(CultureInfo.InvariantCulture));
+        }
+
+        internal void ValidateResponseFromBroker(MsalTokenResponse msalTokenResponse)
+        {
+            _authenticationRequestParameters.RequestContext.Logger.Info(LogMessages.CheckMsalTokenResponseReturnedFromBroker);
+            if (msalTokenResponse.AccessToken != null)
+            {
+                _authenticationRequestParameters.RequestContext.Logger.Info(
+                    LogMessages.BrokerResponseContainsAccessToken +
+                    msalTokenResponse.AccessToken.Count());
+                return;
+            }
+            else if (msalTokenResponse.Error != null)
+            {
+                _authenticationRequestParameters.RequestContext.Logger.Info(
+                    LogMessages.ErrorReturnedInBrokerResponse(msalTokenResponse.Error));
+                throw new MsalServiceException(msalTokenResponse.Error, MsalErrorMessage.BrokerResponseError + msalTokenResponse.ErrorDescription);
```

Comment:
Theoertically we should have a new exception of type `MsalBrokerException`. What are the Android folks doing?

Reply:
no, they only throw a client exception. should still add a new exception type?